react-native/src/main/scala/hello/world/PlayAudio.scala

```diff
@@ -0,0 +1,85 @@
+package hello.world
+
+import org.scalajs.dom.console
+import slinky.core.FunctionalComponent
+import slinky.core.annotations.react
+import slinky.core.facade.Hooks.{useEffect, useState}
+import slinky.native.View
+import typings.antDesignReactNative.components._
+import typings.antDesignReactNative.{antDesignReactNativeStrings => antdStrings}
+import typings.expoAv.aVMod.{AVPlaybackStatus, AVPlaybackStatusToSet}
+import typings.expoAv.mod.Audio.Sound
+import typings.expoAv.anon.{AndroidImplementation, DidJustFinish, Headers}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+
+@react object PlayAudio {
+
+  type Props = Unit
+
+  val component = FunctionalComponent[Props] { _ =>
+    val uri: String = "https://upload.wikimedia.org/wikipedia/commons/b/b2/Funky_bassline_F7-Bb7.ogg"
+    val (status, updateStatus) = useState(None: Option[AVPlaybackStatus])
+    val (isPlayButtonDisabled, updateIsPlayButtonDisabled) = useState(false)
+    val (isPauseButtonDisabled, updateIsPauseButtonDisabled) = useState(true)
+
+    val soundObject = new Sound
+
+    useEffect(
+      () => {
+        soundObject
+          .loadAsync(
+            Headers(uri = uri),
+            AVPlaybackStatusToSet().setShouldPlay(false) // optional, but you can set a bunch of stuff here
+          )
+        soundObject.setOnPlaybackStatusUpdate {
+          status =>
+
+            Option(status.asInstanceOf[DidJustFinish]) match {
+              case Some(obj) =>
+                //@todo this block causes [[Unhandled promise rejection: org.scalajs.linker.runtime.UndefinedBehaviorError: java.lang.ClassCastException: undefined is not an instance of java.lang.Boolean]]
```

Comment:
`AVPlaybackStatus` can be one of two types, and you can distinguish which one based on `isLoaded`

```scala

def checkStatus(s: AVPlaybackStatus): Either[AndroidImplementation, DidJustFinish] = 
  if (s.asInstanceOf[js.Dynamic].isLoaded) Right(s.asInstanceOf[DidJustFinish] else Left(s.asInstanceOf[AndroidImplementation])) 
```

So that exception you're seeing makes sense, because the object you're getting just indicates that the file is loading and the fields youre trying to access arent set, and Scala.js enforces that they are the current type in dev mode

Reply:
Yes, I arrived less elegantly at the same solution, see https://github.com/ScalablyTyped/SlinkyDemos/pull/32/commits/629f980281f420024c19c519f607a867b0eb30fd!

Nevertheless there is a general issue that I have not been able to understand and overcome: any state update (or even `console.log`) placed inside `setOnPlaybackStatusUpdate` seems to cause the unloading of the `soundObject`.

So I reduced the demo to show only how to use the `didJustFinish` flag in a meaningful and I think solid way (but inefficient, always loading before playing). Maybe somebody in the future will be able to go further with pause, stop and so forth.

Reply:
Just found an interesting example for future reference https://blog.jscrambler.com/how-to-build-an-audio-player-in-react-native/

Reply:
It's probably fine, I think the demo is good now anyway. If anybody wants to use this they might have to read docs and debug a bit, but it's still more help now than there was before :)